File: services/analytics_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func, desc
from db.models import Car, Series, Subseries
from core.config import settings

class AnalyticsService:
    @staticmethod
    def get_collection_intelligence(db: Session):
        # 1. Total Models
        total_models = db.query(Car).count()
        
        # 2. Main Series Breakdown
        series_counts = db.query(Series.name, func.count(Car.serial_number)).\
            join(Subseries, Series.id == Subseries.series_id).\
            join(Car, Subseries.id == Car.subseries_id).\
            group_by(Series.name).all()
        main_series_breakdown = {name: count for name, count in series_counts}
        
        # 3. Collection Goals
        # Assuming a target milestone of 500 or 1000. 
        # In a real app, this might be configurable.
        target_milestone = 500 if total_models < 500 else 1000
        progress_percentage = (total_models / target_milestone) * 100 if target_milestone > 0 else 0
        
        collection_goals = {
            "target": target_milestone,
            "current": total_models,
            "progress_percentage": progress_percentage
        }
        
        # 4. Collection Insights
        # Top 5 Subseries
        top_subseries_data = db.query(Subseries.name, func.count(Car.serial_number)).\
            join(Car, Subseries.id == Car.subseries_id).\
            group_by(Subseries.name).\
            order_by(desc(func.count(Car.serial_number))).\
            limit(5).all()
        top_subseries = {name: count for name, count in top_subseries_data}
        
        # Diversity Score: (Unique Subseries / Total Potential Subseries in DB)
        total_subseries_defined = db.query(Subseries).count()
        unique_subseries_owned = db.query(Car.subseries_id).distinct().count()
        diversity_score = (unique_subseries_owned / total_subseries_defined * 100) if total_subseries_defined > 0 else 0
        
        collection_insights = {
            "top_subseries": top_subseries,
            "diversity_score": round(diversity_score, 1)
        }
        
        # 5. Recent Additions
        recent_cars = db.query(Car).join(Subseries).join(Series).\
            order_by(desc(Car.serial_number)).limit(10).all()
            
        recent_additions = []
        for car in recent_cars:
            recent_additions.append({
                "S.No": car.serial_number,
                "Model Name": car.model_name,
                "Series": car.subseries.name if car.subseries else "Unknown",
                "Main Series": car.subseries.series.name if car.subseries and car.subseries.series else "Unknown"
            })
            
        return {
            "total_models": total_models,
            "main_series_breakdown": main_series_breakdown,
            "collection_goals": collection_goals,
            "collection_insights": collection_insights,
            "recent_additions": recent_additions
        }

File: services/analyticsService.js
const Model = require('../models/Model');
const Subseries = require('../models/Subseries');

class AnalyticsService {
  static async getCollectionIntelligence() {
    // 1. Total Models
    const totalModels = await Model.countDocuments();

    // 2. Main Series Breakdown
    const models = await Model.find()
      .populate('metadata.series')
      .populate('metadata.subseries')
      .lean();

    const seriesCounts = {};
    const subseriesCounts = {};

    for (const m of models) {
      const sName = m.metadata.series ? m.metadata.series.name : "Unknown";
      seriesCounts[sName] = (seriesCounts[sName] || 0) + 1;

      const subName = m.metadata.subseries ? m.metadata.subseries.name : "Unknown";
      subseriesCounts[subName] = (subseriesCounts[subName] || 0) + 1;
    }

    // 3. Collection Goals
    const targetMilestone = totalModels < 500 ? 500 : 1000;
    const progressPercentage = targetMilestone > 0 ? (totalModels / targetMilestone) * 100 : 0;

    const collectionGoals = {
      target: targetMilestone,
      current: totalModels,
      progress_percentage: progressPercentage
    };

    // 4. Collection Insights
    const sortedSubseries = Object.entries(subseriesCounts)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 5);
    const topSubseries = Object.fromEntries(sortedSubseries);

    const totalSubseriesDefined = await Subseries.countDocuments();
    const uniqueSubseriesOwned = Object.keys(subseriesCounts).length;
    const diversityScore = totalSubseriesDefined > 0 ? (uniqueSubseriesOwned / totalSubseriesDefined * 100) : 0;

    const collectionInsights = {
      top_subseries: topSubseries,
      diversity_score: Number(diversityScore.toFixed(1))
    };

    // 5. Recent Additions
    const recentModels = await Model.find()
      .populate('metadata.series')
      .populate('metadata.subseries')
      .sort({ serial_number: -1 })
      .limit(10)
      .lean();

    const recentAdditions = recentModels.map(model => ({
      "S.No": model.serial_number,
      "Model Name": model.model_name,
      "Series": model.metadata.subseries ? model.metadata.subseries.name : "Unknown",
      "Main Series": model.metadata.series ? model.metadata.series.name : "Unknown",
      "Brand": model.metadata.brand || "Hot Wheels"
    }));

    return {
      total_models: totalModels,
      main_series_breakdown: seriesCounts,
      collection_goals: collectionGoals,
      collection_insights: collectionInsights,
      recent_additions: recentAdditions
    };
  }
}

module.exports = AnalyticsService;
